orchagent/dtelorch.cpp

```diff
@@ -0,0 +1,1384 @@
+{
+    SWSS_LOG_ENTER();
+
+    if (m_dtelEventTable.find(event) == m_dtelEventTable.end())
+    {
+        return false;
+    }
+
+    return true;
+}
+
+bool DTelOrch::getEventOid(const string& event, sai_object_id_t& oid)
+{
+    SWSS_LOG_ENTER();
+
+    if (!isEventConfigured(event))
+    {
+        SWSS_LOG_ERROR("DTEL ERROR: Event not configured %s", event.c_str());
+        return false;
+    }
+
+    oid = m_dtelEventTable[event].eventOid;
+
+    return true;
+}
+
+void DTelOrch::addEvent(const string& event, const sai_object_id_t& event_oid, const string& report_session_id)
+{
+    if (isEventConfigured(event))
+    {
+        SWSS_LOG_ERROR("DTEL ERROR: Event is already configured %s", event.c_str());
+        return;
+    }
+
+    DTelEventEntry event_entry;
+    event_entry.eventOid = event_oid;
+    event_entry.reportSessionId = report_session_id;
+    m_dtelEventTable[event] = event_entry;
+
+    increaseReportSessionRefCount(report_session_id);
+}
+
+void DTelOrch::removeEvent(const string& event)
+{
+    if (!isEventConfigured(event))
+    {
+        SWSS_LOG_ERROR("DTEL ERROR: Event not configured %s", event.c_str());
+        return;
+    }
+
+    string& report_session_id = m_dtelEventTable[event].reportSessionId;
+
+    decreaseReportSessionRefCount(report_session_id);
+
+    m_dtelEventTable.erase(event);
+}
+
+sai_status_t DTelOrch::updateSinkPortList()
+{
+    sai_attribute_t attr;
+    attr.id = SAI_DTEL_ATTR_SINK_PORT_LIST;
+    sai_status_t status = SAI_STATUS_SUCCESS;
+
+    vector<sai_object_id_t> port_list;
+
+    for (auto it = sinkPortList.begin(); it != sinkPortList.end(); it++)
+    {
+        if (it->second != 0)
+        {
+            port_list.push_back(it->second);
+        }
+    }
+
+    attr.value.objlist.count = (uint32_t)port_list.size();
+    if (port_list.size() == 0)
+    {
+        attr.value.objlist.list = {};   
+    } else {
+        attr.value.objlist.list = port_list.data();
+    }
+    
+    status = sai_dtel_api->set_dtel_attribute(dtelId, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("DTEL ERROR: Failed to set INT sink port list");
+        return status;
+    }
+
+    return status;
+}
+
+bool DTelOrch::addSinkPortToCache(const string &port_alias, const sai_object_id_t port_id)
+{
+    if (sinkPortList.find(port_alias) == sinkPortList.end())
+    {
+        return false;
+    }
+
+    sinkPortList[port_alias] = port_id;
+    return true;
+}
+
+bool DTelOrch::removeSinkPortFromCache(const string &port_alias)
+{
+    if (sinkPortList.find(port_alias) == sinkPortList.end())
+    {
+        return false;
+    }
+
+    sinkPortList[port_alias] = 0;
+    return true;
+}
+
+void DTelOrch::update(SubjectType type, void *cntx)
+{
+    sai_status_t status;
+
+    if (type != SUBJECT_TYPE_PORT_CHANGE)
+    {
+        return;
+    }
+
+    PortUpdate *update = static_cast<PortUpdate *>(cntx);
+
+    /* Check if sink ports need to be updated */
+    if (update->add)
```

Comment:
Hi,
Is this intended to handle the dynamic port breakup case. For example Ethernet128 is configured in the sink port link but not created at startup. Later when the port is created you want to download it to sink port link in hardware. If so I think the logic here doesn't work. From DTelOrch::doDtelTableTask, I see that if port configured in the sink port link doesn't exist, it is not event add into the cache, as a result here the update will not update the newly created port to hardware.
If you are not handling the dynamic port breakup, then all the ports are created at startup, then you just need to wait for the initdone event in APPDB before processing any message, just like what AclOrch::doTask does. In this way you don't need to handle the port update event. 
In short, if here the logic is for handling dynamic port breakup(I am not sure it's already implementd or will be there in future sonic release), then the logic is flawy. If here the log is not for dynamic port breakup, then the logic here seems needless.

Reply:
The case we are handling here is the dynamic port breakup case. I have modified the code to not return an error is the port is not present when the sink ports are configured. The configuration is cached and later played to syncd when the port is added.

Reply:
I agree with dproc the update->add logic is flawed. sink port list should be explicitly specified by user, the orchagent cannot arbitrarily add any breakout port into sink port list.

Reply:
Sink port is explicitly specified by the user and not arbitrarily added on update->add. We check to see if a prior configuration exists for the newly added port and only then push the config down.

Reply:
this is an interesting design. one can argue that if we user add sink ports to config db, all sink ports (physical ports) should exist in the system. If the sink port does not corresponds to a physical port, the configuration should reject that sink port. basically, you cannot configure a sink port without mapping it to a physical port. In this case, whenever you do break out new port will be created and those new ports will not in the sink port list and you won't have above logic. Then after break out, user can then explicitly add the physical port to the sink port list.

Reply:
That was exactly how we had designed this originally. Ports should be created before sink port config. But, as part of design review there were several comments to handle the dynamic port OID change. So, we added this logic. This, in any case, future-proofs the code.

Reply:
it is arguable since if the physical port is never created, then we have mis-match between the config db and the actual asic configuation. I'll defer this to future discussion. For now, we can keep as it is.